### DISCUSSION TITLE
Additional Distance Metrics 🦀

### DISCUSSION THREAD

**Discussion by tusharhqq** [community/community#51712] [Programming Help]
### Body

below is the current implementations of three distance metrics, cosine, dot and euclidian. I want to add 2 additional distance metrics according to this project proposal. 
For users of Qdrant, it would be useful to have the ability to calculate distances using different metrics in addition to the current Euclidean, Cosine, and Dot Product options. These additional metrics, such as Jaccard and Hamming distance, can be useful in specific use cases and provide more accurate results depending on the data and the task.

### Scope

- Implement Jaccard and Hamming distance as new distance metrics options in Qdrant.
- Provide SIMD (Single Instruction, Multiple Data) and Quantized options for these new metrics to improve performance and reduce memory usage.
- Update the documentation to include these new options.

### Motivation

- Jaccard distance is a commonly used metric in information retrieval and text mining and is particularly useful for comparing the similarity of sets of items.
- Hamming distance is commonly used in coding theory and is particularly useful for comparing the similarity of binary strings such as DNA sequences.
- SIMD and Quantized options can improve performance and reduce memory usage, making Qdrant more efficient when working with large datasets.

### Requirements

This feature would require a moderate level of effort as it involves adding new distance metrics to the existing functionality of Qdrant. Knowledge of Rust is necessary. Additionally, knowledge of SIMD and quantization techniques and experience with measuring and optimizing code performance would be beneficial for implementing the SIMD and Quantized options.

use std::arch::x86_64::*;

use crate::data_types::vectors::VectorElementType;
use crate::types::ScoreType;

#[target_feature(enable = "avx")]
#[target_feature(enable = "fma")]
unsafe fn hsum256_ps_avx(x: __m256) -> f32 {
    let x128: __m128 = _mm_add_ps(_mm256_extractf128_ps(x, 1), _mm256_castps256_ps128(x));
    let x64: __m128 = _mm_add_ps(x128, _mm_movehl_ps(x128, x128));
    let x32: __m128 = _mm_add_ss(x64, _mm_shuffle_ps(x64, x64, 0x55));
    _mm_cvtss_f32(x32)
}

#[target_feature(enable = "avx")]
#[target_feature(enable = "fma")]
pub(crate) unsafe fn euclid_similarity_avx(
    v1: &[VectorElementType],
    v2: &[VectorElementType],
) -> ScoreType {
    let n = v1.len();
    let m = n - (n % 32);
    let mut ptr1: *const f32 = v1.as_ptr();
    let mut ptr2: *const f32 = v2.as_ptr();
    let mut sum256_1: __m256 = _mm256_setzero_ps();
    let mut sum256_2: __m256 = _mm256_setzero_ps();
    let mut sum256_3: __m256 = _mm256_setzero_ps();
    let mut sum256_4: __m256 = _mm256_setzero_ps();
    let mut i: usize = 0;
    while i < m {
        let sub256_1: __m256 =
            _mm256_sub_ps(_mm256_loadu_ps(ptr1.add(0)), _mm256_loadu_ps(ptr2.add(0)));
        sum256_1 = _mm256_fmadd_ps(sub256_1, sub256_1, sum256_1);

        let sub256_2: __m256 =
            _mm256_sub_ps(_mm256_loadu_ps(ptr1.add(8)), _mm256_loadu_ps(ptr2.add(8)));
        sum256_2 = _mm256_fmadd_ps(sub256_2, sub256_2, sum256_2);

        let sub256_3: __m256 =
            _mm256_sub_ps(_mm256_loadu_ps(ptr1.add(16)), _mm256_loadu_ps(ptr2.add(16)));
        sum256_3 = _mm256_fmadd_ps(sub256_3, sub256_3, sum256_3);

        let sub256_4: __m256 =
            _mm256_sub_ps(_mm256_loadu_ps(ptr1.add(24)), _mm256_loadu_ps(ptr2.add(24)));
        sum256_4 = _mm256_fmadd_ps(sub256_4, sub256_4, sum256_4);

        ptr1 = ptr1.add(32);
        ptr2 = ptr2.add(32);
        i += 32;
    }

    let mut result = hsum256_ps_avx(sum256_1)
        + hsum256_ps_avx(sum256_2)
        + hsum256_ps_avx(sum256_3)
        + hsum256_ps_avx(sum256_4);
    for i in 0..n - m {
        result += (*ptr1.add(i) - *ptr2.add(i)).powi(2);
    }
    -result
}

#[target_feature(enable = "avx")]
#[target_feature(enable = "fma")]
pub(crate) unsafe fn cosine_preprocess_avx(
    vector: &[VectorElementType],
) -> Option<Vec<VectorElementType>> {
    let n = vector.len();
    let m = n - (n % 32);
    let mut ptr: *const f32 = vector.as_ptr();
    let mut sum256_1: __m256 = _mm256_setzero_ps();
    let mut sum256_2: __m256 = _mm256_setzero_ps();
    let mut sum256_3: __m256 = _mm256_setzero_ps();
    let mut sum256_4: __m256 = _mm256_setzero_ps();
    let mut i: usize = 0;
    while i < m {
        let m256_1 = _mm256_loadu_ps(ptr);
        sum256_1 = _mm256_fmadd_ps(m256_1, m256_1, sum256_1);

        let m256_2 = _mm256_loadu_ps(ptr.add(8));
        sum256_2 = _mm256_fmadd_ps(m256_2, m256_2, sum256_2);

        let m256_3 = _mm256_loadu_ps(ptr.add(16));
        sum256_3 = _mm256_fmadd_ps(m256_3, m256_3, sum256_3);

        let m256_4 = _mm256_loadu_ps(ptr.add(24));
        sum256_4 = _mm256_fmadd_ps(m256_4, m256_4, sum256_4);

        ptr = ptr.add(32);
        i += 32;
    }

    let mut length = hsum256_ps_avx(sum256_1)
        + hsum256_ps_avx(sum256_2)
        + hsum256_ps_avx(sum256_3)
        + hsum256_ps_avx(sum256_4);
    for i in 0..n - m {
        length += (*ptr.add(i)).powi(2);
    }
    if length < f32::EPSILON {
        return None;
    }
    length = length.sqrt();
    Some(vector.iter().map(|x| x / length).collect())
}

#[target_feature(enable = "avx")]
#[target_feature(enable = "fma")]
pub(crate) unsafe fn dot_similarity_avx(
    v1: &[VectorElementType],
    v2: &[VectorElementType],
) -> ScoreType {
    let n = v1.len();
    let m = n - (n % 32);
    let mut ptr1: *const f32 = v1.as_ptr();
    let mut ptr2: *const f32 = v2.as_ptr();
    let mut sum256_1: __m256 = _mm256_setzero_ps();
    let mut sum256_2: __m256 = _mm256_setzero_ps();
    let mut sum256_3: __m256 = _mm256_setzero_ps();
    let mut sum256_4: __m256 = _mm256_setzero_ps();
    let mut i: usize = 0;
    while i < m {
        sum256_1 = _mm256_fmadd_ps(_mm256_loadu_ps(ptr1), _mm256_loadu_ps(ptr2), sum256_1);
        sum256_2 = _mm256_fmadd_ps(
            _mm256_loadu_ps(ptr1.add(8)),
            _mm256_loadu_ps(ptr2.add(8)),
            sum256_2,
        );
        sum256_3 = _mm256_fmadd_ps(
            _mm256_loadu_ps(ptr1.add(16)),
            _mm256_loadu_ps(ptr2.add(16)),
            sum256_3,
        );
        sum256_4 = _mm256_fmadd_ps(
            _mm256_loadu_ps(ptr1.add(24)),
            _mm256_loadu_ps(ptr2.add(24)),
            sum256_4,
        );

        ptr1 = ptr1.add(32);
        ptr2 = ptr2.add(32);
        i += 32;
    }

    let mut result = hsum256_ps_avx(sum256_1)
        + hsum256_ps_avx(sum256_2)
        + hsum256_ps_avx(sum256_3)
        + hsum256_ps_avx(sum256_4);

    for i in 0..n - m {
        result += (*ptr1.add(i)) * (*ptr2.add(i));
    }
    result
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_spaces_avx() {
        use super::*;
        use crate::spaces::simple::*;

        if is_x86_feature_detected!("avx") && is_x86_feature_detected!("fma") {
            let v1: Vec<f32> = vec![
                10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
                10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
                10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
                10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
                26., 27., 28., 29., 30., 31.,
            ];
            let v2: Vec<f32> = vec![
                40., 41., 42., 43., 44., 45., 46., 47., 48., 49., 50., 51., 52., 53., 54., 55.,
                10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
                10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
                10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
                56., 57., 58., 59., 60., 61.,
            ];

            let euclid_simd = unsafe { euclid_similarity_avx(&v1, &v2) };
            let euclid = euclid_similarity(&v1, &v2);
            assert_eq!(euclid_simd, euclid);

            let dot_simd = unsafe { dot_similarity_avx(&v1, &v2) };
            let dot = dot_similarity(&v1, &v2);
            assert_eq!(dot_simd, dot);

            let cosine_simd = unsafe { cosine_preprocess_avx(&v1) };
            let cosine = cosine_preprocess(&v1);
            assert_eq!(cosine_simd, cosine);
        } else {
            println!("avx test skipped");
        }
    }
}

For implementing these distance metric's to this given code. 

Define the new distance metrics: Jaccard and Hamming distance. These can be implemented as functions that take two input vectors (e.g., arrays, slices) and return the distance between them based on the Jaccard or Hamming distance formula.

Add the new distance metrics as options to the existing distance_metric enum in qdrant_controller::config::IndexConfig. This will allow users to specify which distance metric they want to use when creating a new index.

Update the calculate_distance function in qdrant_controller::index_controller::vector_storage::VectorStorage to support the new distance metrics. This function should take the input vectors and the selected distance metric, and return the distance between them.

Implement SIMD and Quantized options for the new distance metrics. SIMD can be used to perform vector operations in parallel, which can significantly improve performance. Quantization can be used to reduce the memory usage by storing the vector data in a compressed format. These options can be implemented as flags in qdrant_controller::config::IndexConfig.

Measure the performance of the new distance metrics with and without SIMD and Quantization using benchmarks. This will allow us to evaluate the effectiveness of these optimizations and fine-tune the implementation if necessary.

Update the documentation to include the new distance metrics and the SIMD and Quantized options. This should include examples of how to use these options in Qdrant and guidance on when to use each distance metric depending on the nature of the data and the task.

Implementing Jaccard and Hamming distance as new distance metrics options in Qdrant can be achieved by modifying the calc_distance function in the vector_storage module. Here's an example implementation that adds Jaccard and Hamming distance options to the function:

pub fn calc_distance(
    &self,
    vec1: &[f32],
    vec2: &[f32],
    metric_type: DistanceMetricType,
) -> Result<f32> {
    match metric_type {
        DistanceMetricType::Euclidean => {
            // implementation of Euclidean distance
        },
        DistanceMetricType::Cosine => {
            // implementation of Cosine distance
        },
        DistanceMetricType::DotProduct => {
            // implementation of Dot Product distance
        },
        DistanceMetricType::Jaccard => {
            // implementation of Jaccard distance
            if vec1.len() != vec2.len() {
                return Err(Error::InvalidArgument(
                    "Vectors must have same length for Jaccard distance".to_string(),
                ));
            }
            let mut intersect = 0;
            let mut union = 0;
            for i in 0..vec1.len() {
                if vec1[i] > 0.0 && vec2[i] > 0.0 {
                    intersect += 1;
                }
                if vec1[i] > 0.0 || vec2[i] > 0.0 {
                    union += 1;
                }
            }
            Ok(1.0 - (intersect as f32 / union as f32))
        },
        DistanceMetricType::Hamming => {
            // implementation of Hamming distance
            if vec1.len() != vec2.len() {
                return Err(Error::InvalidArgument(
                    "Vectors must have same length for Hamming distance".to_string(),
                ));
            }
            let mut dist = 0;
            for i in 0..vec1.len() {
                if vec1[i] != vec2[i] {
                    dist += 1;
                }
            }
            Ok(dist as f32)
        },
    }
}
In this implementation, we first check the metric_type parameter to determine which distance metric to calculate. For Jaccard and Hamming distance, we check that the two vectors have the same length, since these distance metrics are only defined for vectors of the same length.

For Jaccard distance, we calculate the intersection and union of the two sets represented by the binary vectors vec1 and vec2. We then return 1 minus the Jaccard similarity, which is equal to the Jaccard distance.

For Hamming distance, we simply count the number of positions in vec1 and vec2 where the two vectors differ, and return the count as the Hamming distance.

To add SIMD and quantization options to these new distance metrics, we would need to modify the implementation of the distance calculation to take advantage of these techniques. This would likely involve using special instructions provided by the CPU for SIMD operations, and quantizing the input vectors to reduce the amount of memory required to store them. Additionally, we would need to modify the vector_storage module to allow the use of these new options.

can someone go through it and provide there valuable feedback on  this. 
 
### I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

- [X] Yes

**Comment by Bullin112** [comment]
To implement Jaccard and Hamming distance as new distance metrics options in Qdrant, you can start by creating functions that take two sets of items or two binary strings, respectively, as inputs and return the corresponding distance metric as the output.

For Jaccard distance, the function can calculate the size of the intersection and union of the two sets and return their ratio as the distance metric. Here is an example implementation in Rust:
`fn jaccard_distance(set1: &HashSet<&str>, set2: &HashSet<&str>) -> f32 {
    let intersection_size = set1.intersection(set2).count();
    let union_size = set1.union(set2).count();
    1.0 - (intersection_size as f32 / union_size as f32)
}
`
For Hamming distance, the function can compare the binary strings bit by bit and count the number of differences between them. Here is an example implementation in Rust:
`fn hamming_distance(str1: &str, str2: &str) -> u32 {
    str1.chars()
        .zip(str2.chars())
        .filter(|&(a, b)| a != b)
        .count() as u32
}
`
Once you have the distance metric functions implemented, you can create SIMD and quantized options for them to improve performance and reduce memory usage. You can use the same techniques as in the existing implementations of the other distance metrics.

Finally, you can update the documentation to include these new options and provide examples of how to use them. You can also provide guidance on when to use each distance metric depending on the data and the task.

Also get help for [calculations from this site](http://realitycheckcalculator.com/).

**Reply by tusharhqq** [reply]
chatgpt
